backends/gstreamer/src/source.rs

```diff
@@ -38,113 +28,183 @@ mod imp {
+            // the source needs to either:
+            //
+            // 1. be an http, mms, etc. scheme
+            // 2. report that it is "bandwidth limited".
+            //
+            // 1. is not straightforward because we are using a servosrc scheme for now.
+            // This may change in the future if we end up handling http/https/data
+            // URIs, which is what WebKit does.
+            //
+            // For 2. we need to make servosrc handle the scheduling properties query
+            // to report that it "is bandwidth limited".
+            let ret = match query.view_mut() {
+                gst::QueryView::Scheduling(ref mut q) => {
+                    let flags =
+                        gst::SchedulingFlags::SEQUENTIAL | gst::SchedulingFlags::BANDWIDTH_LIMITED;
+                    q.set(flags, 1, -1, 0);
+                    q.add_scheduling_modes(&[gst::PadMode::Push]);
+                    true
+                }
+                _ => false,
+            };
+            let ret = if !ret {
+                pad.query_default(None::<&gst::Pad>, query)
+            } else {
+                true
+            };
+
+            if ret {
+                gst_log!(self.cat, obj: pad, "Handled query {:?}", query);
+            } else {
+                gst_log!(self.cat, obj: pad, "Didn't handle query {:?}", query);
+            }
+            ret
         }
+    }
 
-        pub fn get_type() -> glib::Type {
-            static ONCE: Once = ONCE_INIT;
-            static mut TYPE: glib::Type = glib::Type::Invalid;
+    // Basic declaration of our type for the GObject type system
+    impl ObjectSubclass for ServoSrc {
+        const NAME: &'static str = "ServoSrc";
+        type ParentType = gst::Bin;
+        type Instance = gst::subclass::ElementInstanceStruct<Self>;
+        type Class = subclass::simple::ClassStruct<Self>;
 
-            ONCE.call_once(|| {
-                let t = register_type(ServoSrcStatic);
-                unsafe {
-                    TYPE = t;
-                }
+        glib_object_subclass!();
+
+        // Called once at the very beginning of instantiation of each instance and
+        // creates the data structure that contains all our state
+        fn new_with_class(klass: &subclass::simple::ClassStruct<Self>) -> Self {
+            let app_src = gst::ElementFactory::make("appsrc", None)
+                .map(|elem| elem.downcast::<gst_app::AppSrc>().unwrap())
+                .expect("Could not create appsrc element");
+
+            let pad_templ = klass.get_pad_template("src").unwrap();
+            let ghost_pad = gst::GhostPad::new_no_target_from_template("src", &pad_templ).unwrap();
+
+            ghost_pad.set_query_function(|pad, parent, query| {
+                ServoSrc::catch_panic_pad_function(
+                    parent,
+                    || false,
+                    |queue, element| queue.query(pad.upcast_ref::<gst::Pad>(), element, query),
```

Comment:
Why is the cast here needed?

Reply:
error[E0308]: mismatched types
   --> backends/gstreamer/src/source.rs:122:50
    |
122 |                     |queue, element| queue.query(pad, element, query),
    |                                                  ^^^ expected struct `gst::Pad`, found struct `gst::GhostPad`
    |
    = note: expected type `&gst::Pad`
               found type `&gst::GhostPad`

Reply:
Ah I see, I'll create an issue for this in GStreamer.

Btw, why do you call it `queue`? That's your source here :)

Reply:
Ah nevermind, this is a problem in your code actually.

```rust
        fn query(
            appsrc.set_property_block(false);
            &self,
            appsrc.set_property_format(gst::Format::Bytes);
            pad: &gst::Pad,
            _element: &gst::Element,
            query: &mut gst::QueryRef,
        ) -> bool {
```

This should be

```rust
        fn query<P: IsA<Pad>>(
            appsrc.set_property_block(false);
            &self,
            appsrc.set_property_format(gst::Format::Bytes);
            pad: &P,
            _element: &gst::Element,
            query: &mut gst::QueryRef,
        ) -> bool {
```

Reply:
oops... it's my mistake... I guess I solved :)